### ISSUE TITLE
Support for foreign key/joins in Spring Data Cloud Spanner

### ISSUE THREAD

**Issue by bowenjin** [GoogleCloudPlatform/spring-cloud-gcp#737]
**Is your feature request related to a problem? Please describe.**
The [documentation](https://cloud.spring.io/spring-cloud-gcp/1.1.x/multi/multi__spring_data_cloud_spanner.html) for Spring Data for Cloud Spanner claims that foreign key constraints are not natively supported in cloud spanner. Yet according to [this documentation](https://cloud.google.com/spanner/docs/foreign-keys/overview#how-to-define-foreign-keys), they are supported. While it seems interleaving is the ideal solution for one-to-many relationships, foreign keys are necessary for many-to-many relationships and sometimes for one-to-one. However it seems to ORM mapping annotations for spanner only supports `@Interleaved`. What's the current suggested solution for mapping @OneToOne and @ManyToMany relationships?

**Describe the solution you'd like**
Adding @OneToOne and @ManyToMany mapping annotations to [this package](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/tree/main/spring-cloud-gcp-data-spanner/src/main/java/com/google/cloud/spring/data/spanner/core/mapping)

**Comment by elefeint** [comment]
Support for full JPA specification is out of scope for the Spring Data Spanner module, but with Spanner introducing support for general-purpose foreign keys in March 2020, it may make sense to support a subset such as `OneToOne` and `OneToMany`.

I will leave this feature request open to assess community interest.

Current workarounds:
* As an alternative project to Spring Data Spanner, we also have a [Hibernate dialect](https://github.com/GoogleCloudPlatform/google-cloud-spanner-hibernate#leverage-cloud-spanner-foreign-key-constraints) that does support full JPA relationship mapping.
* For one-to-one relationships, you can also use `@Embedded` in Spring Data Spanner.
* Manually constructed queries with `@Query` and named queries are always an option for crafting well understood joins.

**Comment by meltsufin** [comment]
You're absolutely right in that our documentation is wrong about Spanner not supporting foreign keys. It was written before foreign keys support was added and had not been updated. We should fix it.

That being said, there is nothing preventing you from using foreign key constraints in your Spanner schema while still using Spring Data Spanner. One place we can add foreign keys support is in our schema generation, but I'm not sure if it would make sense without actually introducing new supported annotations like `@OneToMany`.

Like @elefeint says above, if you're looking for a JPA-like ORM, there is Hibernate support that allows to use all of these annotations with Spring Data JPA + Spanner.

**Comment by bowenjin** [comment]
Ah thank you guys, the Hibernate dialect + Spring Data JPA seems like a suitable alternative. Though it would be awesome if Cloud Spanner Spring Data supported some annotations for foreign keys. Hope this issue request gets some interest!

**Comment by mpeddada1** [comment]
Thanks for the responses and for filing the feature request! We will be leaving this issue open. Please feel free to give it a thumbs up if you would like to see this implemented.

**Comment by 2021H1030044G** [comment]
Can I contribute in the above issue?

**Comment by elefeint** [comment]
@2021H1030044G #874 is a better fit for a new contributor.

**Comment by meltsufin** [comment]
I think the idea is that accessing interleaved tables is more efficient because they're collocated in the database.

cc: @olavloite

**Comment by olavloite** [comment]
> I think the idea is that accessing interleaved tables is more efficient because they're collocated in the database.
> 
> cc: @olavloite

Correct:
1. Rows in interleaved tables are phyiscally co-located with their parent row.
2. Accessing a child or the children of a parent row together with the parent row is more efficient than if the tables are not interleaved. Also; fetching all the children of a specific parent will also be more efficient, as these are also all co-located.

So you should prefer using an interleaved table if you have a many-to-one or one-to-one relationship where you mostly access both tables simultaneously, for example by joining the tables in your queries, or if you always select all the children of a given parent row. This as opposed to using two non-interleaved tables and defining a foreign key constraint on the child table. That would also work, and can be a valid solution if you don't always fetch data from both tables in the same query, or don't mostly select all children of a given parent.
